### DISCUSSION TITLE
简单预训练

### DISCUSSION THREAD

**Discussion by xiaomayi-ant** [jingyaogong/minimind#470] [Q&A]
mac m4 24gb 是否能够进行一个基础简单的预训练？

**Comment by basil-k-aji-dev** [comment]
可以跑，但要知道 MPS 在这个仓库里不是一等公民 —— 24GB 统一内存不是瓶颈，**速度和混合精度**才是。下面都是从代码里读出来的。

### 1. `--device mps` 会落进 "cpu" 分支

`trainer/train_pretrain.py:90`，默认设备只在 cuda 和 cpu 之间二选一，永远不会自动选 mps：

```python
parser.add_argument("--device", type=str, default="cuda:0" if torch.cuda.is_available() else "cpu", ...)
```

所以你必须显式传 `--device mps`。但传了之后，第 121 行：

```python
device_type = "cuda" if "cuda" in args.device else "cpu"
```

`"mps"` 里没有 `"cuda"`，于是 `device_type` 变成 `"cpu"`。接着第 123 行：

```python
autocast_ctx = nullcontext() if device_type == "cpu" else torch.cuda.amp.autocast(dtype=dtype)
```

→ **autocast 被关掉，等于全程 fp32 训练**。张量本身还是会通过 `.to(args.device)` 进 MPS（第 28-29 行、第 135 行），所以能算，只是没有混合精度的加速和省显存。

### 2. 有一个坑：不要传 `--dtype float16`

第 138 行：

```python
scaler = torch.cuda.amp.GradScaler(enabled=(args.dtype == 'float16'))
```

默认 `--dtype bfloat16`（第 91 行），所以 `enabled=False`，这个 CUDA 专用的 scaler 就是个空壳，在 Mac 上没事。**但你一旦传 `--dtype float16`，就会在非 CUDA 环境下启用一个 CUDA GradScaler**。保持默认即可。

### 3. 内存不是问题，时间才是

默认配置（第 97-99 行）是 `hidden_size=768`、`num_hidden_layers=8`、`max_seq_len=340`，数据集默认 `pretrain_t2t_mini.jsonl`（README 里标注 1.2GB）。这个量级的模型 + 优化器状态，即使 fp32 也远远用不满 24GB。

真正的参考系是 README 自己给的：

- 官方实验环境是 `RAM 128GB` + `RTX 3090 (24GB) × 8`（第 218-219 行）
- 单卡 3090 跑 1 epoch 预训练约 **2.31 小时**（第 633 行）

M4 的 GPU 在 fp32、没有 CUDA kernel 的情况下，和单卡 3090 的 bf16 混合精度差距会相当大。README 第 284 行本身也写了这一点：

> 若 `cuda` 不可用，也仍可根据自身设备选择 `CPU` 或 `MPS` 运行，但训练速度与兼容性会有非常大的差异。

### 建议

想验证流程能跑通，而不是真的训出一个可用模型：

```bash
python train_pretrain.py --device mps --num_workers 0 --batch_size 4 --accumulation_steps 2 --epochs 1
```

- `--num_workers 0`：默认是 8（第 92 行），macOS 上多进程 DataLoader 容易出问题，先关掉排除干扰
- `batch_size` 调小，先确认一个 step 能跑完、loss 在降，再往上加

说明一下：以上是读代码得出的，我手上没有 M4 机器，没有实测过吞吐。@jingyaogong 如果有人实测过 MPS 的数据，欢迎补充。

基于当前 main（`cc312c1`）。
